### PR TITLE
DCOS Strict security support

### PR DESCRIPTION
**Pull request by gisjedi**
Greetings, Olivier! I was very happy to find your package and it has helped me immensely in migrating a Mesos framework that depended on the v0 protobuf scheduler API.

We had a requirement to support Mesos running within DCOS EE under Strict security posture. Thankfully your v1 Python API was built using requests, so I was able to just implement a custom auth class for request that plugged in quite nicely. I'm presently just packing the built tarball of your package into our software, but would prefer to see it added as part of the published mesoshttp package in PyPI (as a new release obviously).

I've tried to make my changes as unobtrusive as possible in the hope that this would be something you'd accept as an enhancement to the existing codebase. I've done my best to leave existing pure Mesos interaction unchanged and only add the needed bits to traverse the additional DCOS security layer. 

If you have any concerns or requests for improvement, I'm happy to work through that with you.

Cheers,
Jonathan

### CODE REVIEW

File: mesoshttp/client.py [hunk 1]
```diff
@@ -366,10 +386,32 @@ def set_credentials(self, principal, secret):
...
+        :type verify: bool
+        '''
+
+        self.requests_auth = DCOSServiceAuth(service_secret)
+        self.principal = self.requests_auth.principal
+
+        cert_file = 'dcos-ca.crt'
+
+        if not verify:
+            response = requests.get('https://leader.mesos/ca/' + cert_file, verify=False)
```
Comment (osallou) [inline mesoshttp/client.py:408]: this is an hard coded uri "leader.mesos", is this mesos dns related name known by framework ? Don't think so it framework is not a DCOS service itself

File: mesoshttp/core.py [hunk 2]
```diff
@@ -6,8 +6,10 @@ class CoreMesosObject(object):
     Internal class to manage driver
     '''
 
-    def __init__(self, mesos_url, frameworkId, streamId):
+    def __init__(self, mesos_url, frameworkId, streamId, requests_auth, verify):
```
Comment (osallou) [inline mesoshttp/core.py]: to keep compatibility, requests_auth and verify should be optional and have defaults

File: mesoshttp/offers.py [hunk 3]
```diff
@@ -12,8 +12,8 @@ class Offer(CoreMesosObject):
     Wrapper class for Mesos offers
     '''
 
-    def __init__(self, mesos_url, frameworkId, streamId, mesosOffer):
-        CoreMesosObject.__init__(self, mesos_url, frameworkId, streamId)
+    def __init__(self, mesos_url, frameworkId, streamId, mesosOffer, requests_auth, verify):
```
Comment (osallou) [inline mesoshttp/offers.py]: same comment, for compatibility, requests_auth and verify should be optional and have defaults

File: mesoshttp/update.py [hunk 4]
```diff
@@ -12,8 +12,8 @@ class Update(CoreMesosObject):
     This class manages Update message from Mesos master
     '''
 
-    def __init__(self, mesos_url, frameworkId, streamId, mesosUpdate):
-        CoreMesosObject.__init__(self, mesos_url, frameworkId, streamId)
+    def __init__(self, mesos_url, frameworkId, streamId, mesosUpdate, requests_auth, verify):
```
Comment (osallou) [inline mesoshttp/update.py]: same comment for request and verify

File: sample/test.py [hunk 5]
```diff
@@ -31,11 +26,20 @@ def run(self):
     def __init__(self):
         logging.basicConfig()
         self.logger = logging.getLogger(__name__)
-        #signal.signal(signal.SIGINT, signal.SIG_IGN)
         logging.getLogger('mesoshttp').setLevel(logging.DEBUG)
         self.driver = None
-        self.client = MesosClient(mesos_urls=['http://127.0.0.1:5050'])
-        #self.client = MesosClient(mesos_urls=['zk://127.0.0.1:2181/mesos'])
+        # By default use ZK for master detection
+        self.client = MesosClient(mesos_urls=['zk://leader.mesos:2181/mesos'])
```
Comment (osallou) [inline sample/test.py]: you suppose here that leader.mesos is known.... not if service is not in DC/OS

File: sample/test.py [hunk 6]
```diff
@@ -31,11 +26,20 @@ def run(self):
...
+        secret = os.getenv('SERVICE_SECRET')
+        if secret:
+            # We are in Enterprise mode and using service account
+            # so we can rely on DCOS AdminRouter getting us to the Master as we
+            # will have an authentication token
+            self.client = MesosClient(mesos_urls=['https://leader.mesos/mesos'])
```
Comment (osallou) [inline sample/test.py]: same comment for dns

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,12 @@
+FROM python:2
```
Comment (osallou) [inline Dockerfile]: dockerfile should be in sample dir , along the python test as it is for testing only, or named as *dockerfile-test-framework* for example

**Review by osallou — CHANGES_REQUESTED** [review]

**Comment by osallou** [comment]
thanks for your proposal.
I made a few remarks, but I will need some more time to understand all possible impacts on existing frameworks

File: sample/test.py (hunk 6 above)
Comment (gisjedi) [reply sample/test.py]: In this case I'd disagree. Since service account and secrets are only going to be in DCOS, we can count on Mesos DNS being available.

File: sample/test.py (hunk 5 above)
Comment (gisjedi) [reply sample/test.py]: Good call. I'll correct.

File: mesoshttp/update.py (hunk 4 above)
Comment (gisjedi) [reply mesoshttp/update.py]: Makes sense. I'll address this.

File: mesoshttp/offers.py (hunk 3 above)
Comment (gisjedi) [reply mesoshttp/offers.py]: Makes sense. I'll address this.

File: mesoshttp/core.py (hunk 2 above)
Comment (gisjedi) [reply mesoshttp/core.py]: Makes sense. I'll address this.

File: mesoshttp/client.py [hunk 8]
```diff
@@ -574,12 +628,13 @@ def __register(self):
         if self.failover_timeout:
             subscribe['subscribe']['framework_info']['failover_timeout'] = self.failover_timeout
 
-        if self.authenticate:
```
Comment (gisjedi) [inline mesoshttp/client.py:577]: It didn't make sense to me to need `authenticate`, `principal`, and `secret`, when you can simply check them serially for being not `None`. I have no strong opinion in either direction, though. This just seemed like a good, purely internal improvement.

File: mesoshttp/client.py (hunk 1 above)
Comment (gisjedi) [reply mesoshttp/client.py:408]: In DCOS, all tasks running under Mesos have access to the `leader.mesos` address as a means to route to the elected Mesos master. The `set_service_account` function would only be used under DCOS EE, as service accounts aren't even available in DCOS Community (and obviously in pure Mesos deployments), so this seemed like an acceptable approach. 

I included the inline attempt to pull the CA public key up front, only to limit the number of unverified TLS requests to 1. This seemed like the best effort in the case a trust bundle is not provided by the package consumer. The bundle can be provided either as a parameter on `MesosClient` construction or by requests environment variable `REQUESTS_CA_BUNDLE`.

I'm open to other suggestions here.

File: Dockerfile (hunk 7 above)
Comment (gisjedi) [reply Dockerfile]: Makes sense. For the Docker daemon to have access to the actual package source when building the image, it does need to be in the root of the project. I'll rename as requested.

**Review by gisjedi — COMMENTED** [review]
I've created a review with my response to your comments. I'm going to address them as my responses indicate. Thanks for the quick feedback!

File: mesoshttp/client.py (hunk 1 above)
Comment (osallou) [reply mesoshttp/client.py:408]: ok, will check further, but if it makes sense only in DCOS EE and as such makes use of mesos dns, I'm ok with that

File: sample/test.py (hunk 5 above)
Comment (osallou) [reply sample/test.py]: as this is for testing only, could be managed via test params or env variables

File: sample/test.py (hunk 6 above)
Comment (osallou) [reply sample/test.py]: as suggested above with your explanation, this is fine for me

File: mesoshttp/client.py (hunk 8 above)
Comment (osallou) [reply mesoshttp/client.py:577]: correct

File: mesoshttp/client.py [hunk 9]
```diff
@@ -39,13 +42,13 @@ class SchedulerDriver(CoreMesosObject):
         Handler to communicate with scheduler
 
         `MesosClient.SchedulerDriver` instance is available after the
-        SUBSCRIBED event with the subcribed event.
+        SUBSCRIBED event with the subscribed event.
         '''
-        def __init__(self, mesos_url, frameworkId, streamId):
+        def __init__(self, mesos_url, frameworkId, streamId, requests_auth, verify):
```
Comment (osallou) [inline mesoshttp/client.py]: same issue as other init you fixed , for backward compatibility, fields should be optional with defaults when creating a SchedulerDriver object

File: sample/test.py (hunk 5 above)
Comment (gisjedi) [reply sample/test.py]: Good suggestion. I've added the ability to set via environment variable.

**Comment by osallou** [comment]
Looks fine. Will do some testing on my side on mesos community with my framework to check for compatibility and merge if ok

**Review by osallou — APPROVED** [review]

**Comment by osallou** [comment]
@gisjedi , I have merged in develop, wait continue a few testing and merge to master after that for a new release (0.3.0)
First tests looks fine (keeps compatibility with previous code), but I need to do more testing (have a mesos issue I need to fix first, not related to the lib)

**Comment by gisjedi** [comment]
Great news! I'll keep an eye out for the 0.3.0 release.

**Comment by osallou** [comment]
merged and release uploaded to pypi, thanks!
